File: zenoh-jni/src/put.rs
//
// Copyright (c) 2023 ZettaScale Technology
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
//
// Contributors:
//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
//

use crate::errors::{Error, Result};
use crate::key_expr::process_kotlin_key_expr;
use crate::utils::decode_byte_array;

use jni::objects::{JByteArray, JString};
use jni::sys::jint;
use jni::JNIEnv;
use std::sync::Arc;
use zenoh::encoding::Encoding;
use zenoh::internal::EncodingInternals;
use zenoh::key_expr::KeyExpr;
use zenoh::prelude::Wait;
use zenoh::publication::{CongestionControl, Priority};
use zenoh::sample::{QoSBuilderTrait, SampleBuilderTrait, ValueBuilderTrait};
use zenoh::session::Session;

/// Performs a `put` operation in the Zenoh session, propagating eventual errors.
///
/// Parameters:
/// - `env`: A mutable reference to the JNI environment.
/// - `key_expr_ptr`: Raw pointer of a declared [KeyExpr] to use for the put operation. If it is null, then
///     the key_expr_str parameter is used.
/// - `key_expr_str`: String representation of the key expression to be used for the put operation. If the key_expr_ptr
///     is valid, then this parameter won't be considered.
/// - `session`: An [Session] to use for the put operation.
/// - `payload`: The payload to send through the network.
/// - `encoding`: The [Encoding] of the put operation.
/// - `congestion_control`: The [CongestionControl] mechanism specified.
/// - `priority`: The [Priority] mechanism specified.
/// - `attachment`: An optional attachment, encoded into a byte array. May be null.
///
/// Returns:
/// - A `Result` indicating the result of the `get` operation, with an [Error] in case of failure.
///
#[allow(clippy::too_many_arguments)]
pub(crate) fn on_put(
    env: &mut JNIEnv,
    key_expr_ptr: *const KeyExpr<'static>,
    key_expr_str: JString,
    session: &Arc<Session>,
    payload: JByteArray,
    encoding: jint,
    congestion_control: jint,
    priority: jint,
    attachment: JByteArray,
) -> Result<()> {
    let key_expr = unsafe { process_kotlin_key_expr(env, &key_expr_str, key_expr_ptr) }?;
    let payload = decode_byte_array(env, payload)?;
    let encoding = Encoding::new(encoding as u16, None); // TODO: provide schema
    let congestion_control = match decode_congestion_control(congestion_control) {
        Ok(congestion_control) => congestion_control,
        Err(err) => {
            tracing::warn!(
                "Error decoding congestion control: '{}'. Using default...",
                err
            );
            CongestionControl::default()
        }
    };

    let priority = match decode_priority(priority) {
        Ok(priority) => priority,
        Err(err) => {
            tracing::warn!("Error decoding priority: '{}'. Using default...", err);
            Priority::default()
        }
    };

    let mut put_builder = session
        .put(&key_expr, payload)
        .congestion_control(congestion_control)
        .encoding(encoding)
        .priority(priority);

    if !attachment.is_null() {
        let attachment = decode_byte_array(env, attachment)?;
        put_builder = put_builder.attachment(attachment)
    }

    put_builder
        .wait()
        .map(|_| tracing::trace!("Put on '{key_expr}'"))
        .map_err(|err| Error::Session(format!("{err}")))
}

pub(crate) fn decode_priority(priority: jint) -> Result<Priority> {
    match Priority::try_from(priority as u8) {
        Ok(priority) => Ok(priority),
        Err(err) => Err(Error::Session(format!("Error retrieving priority: {err}."))),
    }
}

pub(crate) fn decode_congestion_control(congestion_control: jint) -> Result<CongestionControl> {
    match congestion_control {
        1 => Ok(CongestionControl::Block),
        0 => Ok(CongestionControl::Drop),
        _value => Err(Error::Session(format!(
            "Unknown congestion control '{_value}'."
        ))),
    }
}

File: zenoh-java/src/commonMain/kotlin/io/zenoh/publication/Put.kt
//
// Copyright (c) 2023 ZettaScale Technology
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
//
// Contributors:
//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
//

package io.zenoh.publication

import io.zenoh.Resolvable
import io.zenoh.Session
import io.zenoh.exceptions.ZenohException
import io.zenoh.keyexpr.KeyExpr
import io.zenoh.prelude.Encoding
import io.zenoh.prelude.SampleKind
import io.zenoh.value.Value

/**
 * Put operation.
 *
 * A put puts a [io.zenoh.sample.Sample] into the specified key expression.
 *
 * Example:
 * ```java
 * try (Session session = Session.open()) {
 *     try (KeyExpr keyExpr = KeyExpr.tryFrom("demo/example/zenoh-java-put")) {
 *         String value = "Put from Java!";
 *         session.put(keyExpr, value)
 *             .congestionControl(CongestionControl.BLOCK)
 *             .priority(Priority.REALTIME)
 *             .kind(SampleKind.PUT)
 *             .res();
 *         System.out.println("Putting Data ('" + keyExpr + "': '" + value + "')...");
 *     }
 * }
 * ```
 *
 * This class is an open class for the sake of the [Delete] operation, which is a special case of [Put] operation.
 *
 * @property keyExpr The [KeyExpr] to which the put operation will be performed.
 * @property value The [Value] to put.
 * @property congestionControl The [CongestionControl] to be applied when routing the data.
 * @property priority The [Priority] of zenoh messages.
 * @property kind The [SampleKind] of the sample (put or delete).
 */
open class Put protected constructor(
    val keyExpr: KeyExpr,
    val value: Value,
    val congestionControl: CongestionControl,
    val priority: Priority,
    val kind: SampleKind
) {

    companion object {

        /**
         * Creates a bew [Builder] associated to the specified [session] and [keyExpr].
         *
         * @param session The [Session] from which the put will be performed.
         * @param keyExpr The [KeyExpr] upon which the put will be performed.
         * @param value The [Value] to put.
         * @return A [Put] operation [Builder].
         */
        fun newBuilder(session: Session, keyExpr: KeyExpr, value: Value): Builder {
            return Builder(session, keyExpr, value)
        }
    }

    /**
     * Builder to construct a [Put] operation.
     *
     * @property session The [Session] from which the put operation will be performed.
     * @property keyExpr The [KeyExpr] upon which the put operation will be performed.
     * @property value The [Value] to put.
     * @property congestionControl The [CongestionControl] to be applied when routing the data,
     *  defaults to [CongestionControl.DROP]
     * @property priority The [Priority] of zenoh messages, defaults to [Priority.DATA].
     * @property kind The [SampleKind] of the sample (put or delete), defaults to [SampleKind.PUT].
     * @constructor Create a [Put] builder.
     */
    class Builder internal constructor(
        private val session: Session,
        private val keyExpr: KeyExpr,
        private var value: Value,
        private var congestionControl: CongestionControl = CongestionControl.DROP,
        private var priority: Priority = Priority.DATA,
        private var kind: SampleKind = SampleKind.PUT
    ): Resolvable<Unit> {

        /** Change the [Encoding] of the written data. */
        fun encoding(encoding: Encoding) = apply {
            this.value = Value(value.payload, encoding)
        }

        /** Change the [CongestionControl] to apply when routing the data. */
        fun congestionControl(congestionControl: CongestionControl) =
            apply { this.congestionControl = congestionControl }

        /** Change the [Priority] of the written data. */
        fun priority(priority: Priority) = apply { this.priority = priority }

        /** Change the [SampleKind] of the sample. If set to [SampleKind.DELETE], performs a delete operation. */
        fun kind(kind: SampleKind) = apply { this.kind = kind }

        /** Resolves the put operation. */
        @Throws(ZenohException::class)
        override fun res() {
            val put = Put(keyExpr, value, congestionControl, priority, kind)
            session.run { resolvePut(keyExpr, put) }
        }
    }
}
